src/type_manager.cpp

```diff
@@ -81,20 +84,82 @@ void TypeManager::register_methods(jni::Env& p_env, jni::JObjectArray& method_na
     integer_class.delete_local_ref(p_env);
 }
 
-void TypeManager::register_user_types(jni::Env& p_env, jni::JObjectArray& p_types) {
-#ifdef DEBUG_ENABLED
-    LOG_VERBOSE("Starting to register user types...");
+void TypeManager::create_and_update_scripts(Vector<KtClass*>& classes) {
+    LocalVector<Ref<KotlinScript>> scripts;
+
+#ifdef TOOLS_ENABLED
+    // This tool-only block handles script reloading.
+    // We have to compare the previous scripts to the new ones and create/update/delete accordingly
+
+    HashMap<StringName, Ref<KotlinScript>> script_cache = user_scripts_map;
+    user_scripts.clear();
+    user_scripts_map.clear();
+
+    for (KtClass* kotlin_class : classes) {
+        // First check if the scripts already exist
+        Ref<KotlinScript> ref = script_cache[kotlin_class->registered_class_name];
+        if (!ref.is_null()) {
+            delete ref->kotlin_class;
+            ref->kotlin_class = kotlin_class;
+#ifdef DEV_ENABLED
+            LOG_VERBOSE(vformat("Kotlin Script updated: %s", kotlin_class->registered_class_name));
 #endif
-    for (int i = 0; i < p_types.length(p_env); ++i) {
-        const String& script_path {p_env.from_jstring(static_cast<jni::JString>(p_types.get(p_env, i)))};
-        Ref<KotlinScript> script = ResourceLoader::load(script_path, "KotlinScript");
-        user_scripts.insert(i, script);
-        user_scripts_map[script->get_global_name()] = script;
-#ifdef DEBUG_ENABLED
-        LOG_VERBOSE(vformat("Registered %s user type with index %s.", script_path, i));
+        } else {
+            // Script doesn't exist so we create it.
+            ref.instantiate();
+            ref->kotlin_class = kotlin_class;
+#ifdef DEV_ENABLED
+            LOG_VERBOSE(vformat("Kotlin Script created: %s", kotlin_class->registered_class_name));
 #endif
+        }
+
+        scripts.push_back(ref);
+        script_cache.erase(kotlin_class->registered_class_name);
+    }
+
+    // Only scripts left in the cache are the ones that have been removed or placeholders without associated .kt
+    // We simply delete their kotlin_class if they got one
+    for (KeyValue<StringName, Ref<KotlinScript>> keyValue : script_cache) {
+        Ref<KotlinScript> ref = keyValue.value;
+        if (ref->kotlin_class) { delete ref->kotlin_class; }

```

Comment:
We logged earlier which scripts got created and updated. Shouldn't we also log which ones get deleted?